### ISSUE TITLE
UpdateReportedPropertiesAsync not updating device twin

### ISSUE THREAD

**Issue by amogh79** [Azure/azure-iot-sdk-csharp#340]
- **OS and version used:** Windows 10

- **SDK version used:** Microsoft.Azure.Devices.Client 1.6.1

# Description of the issue:

I am simulating a device for desired properties changed event. I am using UpdateReportedPropertiesAsync method on DeviceClient to update reported properties, I don't get any error but the reported properties for the device are not getting updated. I checked both on IoT hub and retrieving the devicetwin using GetTwinAsync from within the code. We are nearing a release next week and any help on this would be highly appreciated.

# Code sample exhibiting the issue:

# Console log of the issue:

**Comment by invinctus** [comment]
I can also reproduce this it seems to not write properties when you are using arrays.

`var prop = new TwinCollection("{\"collection\":[{\"Id\": 1}]}");`  This wont update the properties
`var prop = new TwinCollection("{\"collection\":{\"Id\": 1}}");` This updates ok.

Is it not possible to have a json array in the twin data?
Should the method not return some indication that it is unable to write these properties?

**Comment by amogh79** [comment]
For me its failing in all cases. Following is the code snippet wherein it fails and doesn't update the reported properties for the device.

`TwinCollection test = new TwinCollection(
                    "{\r\n  \"softwareConfig\": {\r\n    \"softwareName\": \"reg003SE\",\r\n    \"version\": \"2.0.0\",\r\n    \"softwareURL\": \"OneDrive_1_12-21-2017.zip\",\r\n    \"serialNumber\": \"demotestdevice01\",\r\n    \"softwarePackageId\": \"da475a46-405c-4388-bc3e-67e6fae83bba\"\r\n  },\r\n  \"$version\": 6\r\n}");


                await _client.UpdateReportedPropertiesAsync(test);`

**Comment by invinctus** [comment]
@amogh79 you can get yours to work if you omit the "$" in the "$version" property it doesn't seem to like properties that begin with $

**Comment by amogh79** [comment]
Jeez how come I forgot that properties starting with "$" are readonly. Thanks for your inputs @invinctus.
You can use TwinCollection.ClearMetaData() to clean it off readonly properties.

**Comment by CIPop** [comment]
> Jeez how come I forgot that properties starting with "$" are readonly. Thanks for your inputs @invinctus.

@amogh79 does this mean that it worked for you after removing "$"? 

> Should the method not return some indication that it is unable to write these properties?

The service should return an error. If the "$" was the reason you couldn't update properties, I'd like to repurpose this issue to track the lack of error reporting.

> I can also reproduce this it seems to not write properties when you are using arrays.

@invinctus Does it work if you add properties on the collection instead of parsing them from a `string`?
Also, please open a separate issue to track correct array parsing (I'd like to keep this as higher priority for updating properties / getting correct errors).

**Comment by amogh79** [comment]
@CIPop Yes it worked for me after removing '$'. As per device twins's documentation [here](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins) it says some properties are readonly and are set by device/platform. I believe the one's starting with '$' are readonly.

**Comment by invinctus** [comment]
@CIPop the documentation does say arrays are not supported I didn't notice that initially.  The biggest issue really is the lack of error reporting when you try to write something that is not supported.

**Comment by CIPop** [comment]
Thanks @amogh79 and @invinctus ! Glad to see you are unblocked! 
I would like to investigate not getting an error which at this point seems to be a service-side issue.

@simonporter I'm not very familiar with the Service SDK but in general we would expect the server side to report back if any of the data was incorrect. Is there any reason that the server would accept both the array and $ param changes silently?

**Comment by zolvarga** [comment]
@amogh79 @invinctus @CIPop 

If we send invalid names in reported properties service will send back rejection with status 400.
The handling of those error codes were missing in the Device Client code.

I added handler for that, currently in PR. As soon it will be merged I will close this issue.

Best Regards,
Zoltan
